### PR TITLE
Improve GCS UFS

### PR DESCRIPTION
**Pull request by LuQQiu**
This PR improves the GCS UFS similar to our previous improvements to S3A UFS.
(1) move all permission related code to `GCSUnderFileSystem.getPermissionInternal()` and memorize the permission (only get the permission for once and remember it). This change mocks what S3A UFS does.
(2) In `GCSInputStream`, mainly reorganize the codes. Move create `inputStream` related codes in the `openStream()` other than in the constructor. The constructor suppose to not throw exceptions.  Use `openStream()` and `closeStream()` similar to what S3A UFS does.
(3) remove the `UnderFileSystemConfiguration mConf` in `S3AUnderFileSystem` since its super class `BaseUnderFileSystem` already have the member `UnderFileSystemConfiguration mUfsConf`.

### CODE REVIEW

**Comment by gpang** [comment]
@LuQQiu could you provide a description for the PR?

**Comment by LuQQiu** [comment]
@gpang Added the description. We want to apply our improvements on S3A to GCS if applicable. This PR includes all the non big feature GCS improvements. The other bigger improvements like GCS streaming transfer will be given in future PR. I have used `./bin/alluxio runUfsTests --path gs://<bucket>/ -Dfs.gcs.accessKeyId=<access_key> -Dfs.gcs.secretAccessKey=<secret_key` to test GCS UFS changes locally. PTAL, thanks

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java [hunk 1]
```diff
@@ -593,11 +570,7 @@ protected String getRootKey() {
 
   @Override
   protected InputStream openObject(String key, OpenOptions options,
-      RetryPolicy retryPolicy) throws IOException {
-    try {
-      return new S3AInputStream(mBucketName, key, mClient, options.getOffset(), retryPolicy);
-    } catch (AmazonClientException e) {
```
Comment (gpang) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java]: why is this changed?

File: underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSInputStream.java [hunk 2]
```diff
@@ -147,21 +134,55 @@ public int read(byte[] b, int off, int len) throws IOException {
...
+        GSObject object;
+        if (mPos > 0) {
+          object = mClient.getObject(mBucketName, mKey, null, null, null, null, mPos, null);
+        } else {
+          object = mClient.getObject(mBucketName, mKey);
+        }
+        mInputStream = new BufferedInputStream(object.getDataInputStream());
+        return;
+      } catch (ServiceException e) {
+        LOG.warn("Attempt {} to open key {} in bucket {} failed with exception : {}",
```
Comment (gpang) [inline underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSInputStream.java]: should we include `mPos` in the message?

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java (hunk 1 above)
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java]: Because `S3AInputStream` constructor does not throw `AmazonClientException`, i remove the catch of exceptions.

File: underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSInputStream.java (hunk 2 above)
Comment (LuQQiu) [reply underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSInputStream.java]: Added the mPos information

**Comment by LuQQiu** [comment]
Thanks @gpang! 
@calvinjia Could you also take a look at the PR, thanks

File: underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java [hunk 3]
```diff
@@ -313,14 +287,63 @@ protected ObjectStatus getObjectStatus(String key) {
...
+   * effort and will continue with default permissions (no owner, no group, 0700).
+   *
+   * @return the permissions associated with this under storage system
+   */
+  private ObjectPermissions getPermissionsInternal() {
+    // getAccountOwner() can return null even when the account is authenticated.
+    // TODO(chaomin): investigate the root cause why Google cloud service is returning
+    // null StorageOwner.
+    String accountOwnerId = "unknown";
+    String accountOwner = "unknown";
```
Comment (calvinjia) [inline underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java]: These defaults are not consistent with the javadoc?

Comment (LuQQiu) [reply underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java]: Changed the default to `""` to consistent with javadoc

**Comment by calvinjia** [comment]
alluxio-bot, merge this please.

**Comment by calvinjia** [comment]
alluxio-bot, cherry-pick this to branch-2.0 please.
